### ISSUE TITLE
Memory leak leads to OutOfMemoryException in some cases

### ISSUE THREAD

**Issue by Pathoschild** [Pathoschild/SMAPI#685]
Some players report `OutOfMemoryException` errors since the Stardew Valley 1.4.2 release (see timeline below). Once it started happening, previous versions of SMAPI and Stardew Valley were retroactively affected.

## Reports
report | game | SMAPI | CP | log
------ | ---- | ----- | -- | ----
[2019-12-05 hiabara on Nexus](https://www.nexusmods.com/stardewvalley/mods/2400/?tab=posts&jump_to_comment=75428278)<br /><sup>(asset reload issue)</sup> | 1.4.2 | 3.0.1 | 1.10.1 | log [A](https://smapi.io/log/1e60c025ccd8411a9688c1fb8860f5ea) and [B](https://smapi.io/log/X4h6Acfu)
[2019-12-12 Alvaror on Nexus](https://www.nexusmods.com/stardewvalley/mods/2449?tab=posts&jump_to_comment=75629168)<br /><sup>(asset reload issue; on Discord says started c. 8th)</small> | 1.4.2 | 3.0.1 | 1.10.1 | [log](https://smapi.io/log/60ef1f5074e249f28b56198ee69f5c19)
[2019-12-12 Spartan69 on Nexus](https://www.nexusmods.com/stardewvalley/mods/2449?tab=bugs&issue_id=433850)<br /><sup>(asset reload issue)</sup> | 1.4.2 | 3.0.1 |1.10.1 | [log](https://smapi.io/log/rKgRqmyb)
[2019-12-24 snesknees on Nexus](https://www.nexusmods.com/stardewvalley/mods/2400/?tab=posts&jump_to_comment=76003128)<br /><sup>(return to title issue)</sup> | 1.4.3 | 3.0.1 | 1.11 | [log](https://smapi.io/log/6b9fc7ab2dfb44eca71fe18bd0057c49)
[2019-12-28 mihihia on Nexus](https://www.nexusmods.com/stardewvalley/mods/2400/?tab=posts&jump_to_comment=76037563) | 1.4.3 | 3.0.1 | 1.11.1 | [log](https://smapi.io/log/6811e00d738b451bb88d1d534e2a0772)
[2019-12-29 buckminst on Nexus](https://www.nexusmods.com/stardewvalley/mods/1915/?tab=posts&jump_to_comment=76054488&BH=1) | 1.4.3 | 3.0.1 | 1.11.1 | [log](https://smapi.io/log/126acf8febb2401095372af81b085bea)

## Timeline
date | event
---- | -----
2019-11-26 | Stardew Valley 1.4<br />SMAPI 3.0<br />Content Patcher 1.10
2019-12-02 | Stardew Valley 1.4.1<br />SMAPI 3.0.1<br />Content Patcher 1.10.1
2019-12-04 | Stardew Valley 1.4.2
**2019-12-05** | **first report**

**Comment by Pathoschild** [comment]
Tested scenarios:
* **SDV 1.4.3 + SMAPI 3.0.1 + CP 1.11 seasonal pack**  
  Most reports involve Content Patcher packs which change textures often. For example, here I repeatedly went indoors/outdoors to reload seasonal textures until it crashed:
  > ![](https://user-images.githubusercontent.com/230581/71526060-db765600-28a2-11ea-9626-c4bfbd8f3a88.png)

* **SDV 1.4.3 + SMAPI 3.0.1 (no mods)**  
  It also happens with SMAPI but no mods, by repeatedly loading a save and returning to title. For example, here I did that until the game became unbearably slow (but didn't crash outright):
  > ![](https://user-images.githubusercontent.com/230581/71526613-b20af980-28a5-11ea-9aed-30fd96d00952.png)

* **SDV 1.3.36 + SMAPI 2.11.3 (no mods)**  
  It also happens with Stardew Valley 1.3.36 (released January 2019) and SMAPI 2.11.3 (released September 2019), even though no issues were reported until December 2019 (and multiple issues were reported since then).
  > ![](https://user-images.githubusercontent.com/230581/71527289-c1d80d00-28a8-11ea-9cba-a37a4ed13e4c.png)

* **SDV 1.4.3 without SMAPI**  
  It doesn't happen without SMAPI.
  > ![](https://user-images.githubusercontent.com/230581/71526831-9eac5e00-28a6-11ea-802a-e8de51d65719.png)  

This happens even if I strip out SMAPI's content pipeline, so it's not strictly related to asset handling. It doesn't happen if I run SMAPI with _all_ features and APIs stripped out, so it's something in the SMAPI code.

**Comment by Esca-MMC** [comment]
Per my comments in Discord, this seems to be an issue in every build of SDV + SMAPI I've managed to test. I first noticed some memory leakage in October or November of 2019. Task Manager reports ~20MB more memory each time I open the "Load" menu, and ~60MB more each time I load any saved game. The results are similar with or without any mods.

I've tested the following versions and experienced similar leak rates:

* SDV 1.4.3 + SMAPI 3.0.1
* SDV 1.3.36 + SMAPI 2.11.3
* SDV 1.3.36 + SMAPI 2.11-beta
* SDV 1.3.27 + SMAPI 2.6

As you mentioned, this could be a persistent (but previously benign) memory leak or a result of more recent Windows updates. These specific results happened in Windows 10 Pro, Build 18362 (last updated in July 2019).
